File: routa-core/src/main/kotlin/com/phodal/routa/core/pipeline/stages/CrafterExecutionStage.kt
package com.phodal.routa.core.pipeline.stages

import com.phodal.routa.core.coordinator.CoordinationPhase
import com.phodal.routa.core.model.AgentRole
import com.phodal.routa.core.model.AgentStatus
import com.phodal.routa.core.model.CompletionReport
import com.phodal.routa.core.pipeline.PipelineContext
import com.phodal.routa.core.pipeline.PipelineStage
import com.phodal.routa.core.pipeline.RetryPolicy
import com.phodal.routa.core.pipeline.StageResult
import com.phodal.routa.core.runner.OrchestratorPhase
import com.phodal.routa.core.runner.OrchestratorResult
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope

/**
 * **Stage 3: CRAFTER Execution** — Executes all ready tasks with CRAFTER agents.
 *
 * This stage:
 * 1. Asks the coordinator for the next wave of ready tasks
 * 2. Creates CRAFTER agents for each task
 * 3. Builds context (task definition + agent identity) for each CRAFTER
 * 4. Runs CRAFTERs (sequentially or in parallel based on [PipelineContext.parallelCrafters])
 * 5. Ensures completion reports are filed (via tool calls or text parsing fallback)
 * 6. Cleans up provider resources for completed agents
 *
 * ## Outputs (written to [PipelineContext])
 * - [PipelineContext.delegations] — list of (agentId, taskId) pairs executed
 *
 * ## Early termination
 * If no tasks are ready AND all tasks are completed, returns
 * [StageResult.SkipRemaining] with a success result (no verification needed).
 */
class CrafterExecutionStage : PipelineStage {

    override val name = "crafter-execution"
    override val description = "Executes ready tasks with CRAFTER agents"

    /**
     * Retry policy for CRAFTER execution.
     *
     * LLM calls can fail due to network issues or rate limits.
     * We retry up to 2 times with exponential backoff (2s → 4s).
     */
    override val retryPolicy = RetryPolicy(
        maxAttempts = 2,
        baseDelayMs = 2000,
        backoffMultiplier = 2.0,
    )

    override suspend fun execute(context: PipelineContext): StageResult {
        context.waveNumber = context.waveNumber + 1
        context.emitPhase(OrchestratorPhase.WaveStarting(context.waveNumber))

        // Ask the coordinator for the next wave of tasks
        val delegations = context.system.coordinator.executeNextWave()

        if (delegations.isEmpty()) {
            // No tasks to execute — check if everything is already done
            val phase = context.system.coordinator.coordinationState.value.phase
            if (phase == CoordinationPhase.COMPLETED) {
                context.emitPhase(OrchestratorPhase.Completed)
                return StageResult.SkipRemaining(buildSuccessResult(context))
            }
            // Nothing to do but not completed — continue to next stage
            return StageResult.Continue
        }

        context.delegations = delegations

        // Run CRAFTERs — parallel or sequential
        if (context.parallelCrafters && delegations.size > 1) {
            coroutineScope {
                val jobs = delegations.map { (crafterId, taskId) ->
                    async { runSingleCrafter(context, crafterId, taskId) }
                }
                jobs.awaitAll()
            }
        } else {
            for ((crafterId, taskId) in delegations) {
                runSingleCrafter(context, crafterId, taskId)
            }
        }

        return StageResult.Continue
    }

    // ── Single CRAFTER execution ────────────────────────────────────────

    /**
     * Run a single CRAFTER agent with streaming and report handling.
     */
    private suspend fun runSingleCrafter(
        context: PipelineContext,
        crafterId: String,
        taskId: String,
    ) {
        context.ensureActive() // ← cancellation check before each CRAFTER
        context.emitPhase(OrchestratorPhase.CrafterRunning(crafterId, taskId))

        val taskContext = context.system.coordinator.buildAgentContext(crafterId)
            ?: return // No task assigned
        val prompt = injectAgentIdentity(taskContext, crafterId, taskId)

        val crafterOutput = context.provider.runStreaming(
            AgentRole.CRAFTER, crafterId, prompt
        ) { chunk ->
            context.onStreamChunk?.invoke(crafterId, chunk)
        }

        // Ensure the CRAFTER's work is reported
        ensureCrafterReport(context, crafterId, taskId, crafterOutput)

        // Clean up provider resources
        context.provider.cleanup(crafterId)

        context.emitPhase(OrchestratorPhase.CrafterCompleted(crafterId, taskId))
    }

    // ── Report handling ─────────────────────────────────────────────────

    /**
     * If the CRAFTER didn't call `report_to_parent` via tool call,
     * parse the text output and file the report on their behalf.
     */
    private suspend fun ensureCrafterReport(
        context: PipelineContext,
        crafterId: String,
        taskId: String,
        output: String,
    ) {
        val agent = context.system.context.agentStore.get(crafterId) ?: return

        // If agent is already COMPLETED, Koog tool calling handled it
        if (agent.status == AgentStatus.COMPLETED) return

        // Agent didn't call report_to_parent — parse text and report for them
        val report = context.reportParser.parseCrafterReport(crafterId, taskId, output)
        context.system.tools.reportToParent(crafterId, report)
    }

    // ── Helpers ──────────────────────────────────────────────────────────

    /**
     * Inject the agent's identity into the prompt so the LLM knows
     * what values to pass when calling tools like `report_to_parent`.
     */
    private fun injectAgentIdentity(prompt: String, agentId: String, taskId: String): String {
        return buildString {
            appendLine("--- Agent Identity ---")
            appendLine("agent_id: $agentId")
            appendLine("task_id: $taskId")
            appendLine("----------------------")
            appendLine()
            append(prompt)
        }
    }

    private suspend fun buildSuccessResult(context: PipelineContext): OrchestratorResult.Success {
        val summary = context.system.coordinator.getTaskSummary()
        return OrchestratorResult.Success(summary)
    }
}

File: routa-js/src/core/pipeline/stages/crafter-execution-stage.ts
/**
 * CrafterExecutionStage - port of routa-core CrafterExecutionStage.kt
 *
 * Stage 3: CRAFTER Execution - Executes all ready tasks with CRAFTER agents.
 */

import { AgentRole } from "../../models/agent";
import { TaskStatus } from "../../models/task";
import { PipelineStage } from "../pipeline-stage";
import { PipelineContext, emitPhase } from "../pipeline-context";
import { StageResult } from "../stage-result";
import { CoordinationPhase } from "../../coordinator/coordination-state";

/**
 * Stage 3: CRAFTER Execution
 *
 * This stage:
 * 1. Gets the next wave of ready tasks from the coordinator
 * 2. Creates CRAFTER agents for each task
 * 3. Builds context for each CRAFTER
 * 4. Runs CRAFTERs (sequentially or in parallel)
 * 5. Marks tasks as REVIEW_REQUIRED when complete
 */
export class CrafterExecutionStage implements PipelineStage {
  name = "crafter-execution";
  description = "Executes ready tasks with CRAFTER agents";

  async execute(context: PipelineContext): Promise<StageResult> {
    context.waveNumber = context.waveNumber + 1;
    await emitPhase(context, {
      type: "wave_starting",
      waveNumber: context.waveNumber,
    });

    // Get ready tasks
    const readyTasks = await context.context.taskStore.findReadyTasks(
      context.workspaceId
    );

    if (readyTasks.length === 0) {
      // Check if all tasks are completed
      const allTasks = await context.context.taskStore.listByWorkspace(
        context.workspaceId
      );
      if (allTasks.every((t) => t.status === TaskStatus.COMPLETED)) {
        await emitPhase(context, { type: "completed" });
        return StageResult.SkipRemaining({
          type: "success",
          taskSummaries: allTasks.map((t) => t.title),
          planOutput: context.planOutput,
        });
      }
      // Nothing to do but not completed — continue to next stage
      return StageResult.Continue();
    }

    const delegations: Array<[string, string]> = [];

    // Execute each task
    for (const task of readyTasks) {
      // Create CRAFTER agent
      const crafterName = `crafter-${task.title
        .toLowerCase()
        .replace(/[^a-z0-9]+/g, "-")}`;

      const result = await context.context.agentTools.createAgent({
        name: crafterName,
        role: AgentRole.CRAFTER,
        workspaceId: context.workspaceId,
        parentId: context.routaAgentId,
      });

      if (!result.success || !result.data) {
        continue;
      }

      const agentId = (result.data as { agentId: string }).agentId;

      // Build task context
      const taskPrompt = this.buildTaskPrompt(task);

      // Execute CRAFTER
      await context.provider.runStreaming(
        AgentRole.CRAFTER,
        agentId,
        taskPrompt,
        (chunk) => {
          if (context.onStreamChunk) {
            context.onStreamChunk(agentId, chunk);
          }
        }
      );

      // Mark task as REVIEW_REQUIRED
      await context.context.taskStore.save({
        ...task,
        status: TaskStatus.REVIEW_REQUIRED,
      });

      delegations.push([agentId, task.id]);
    }

    context.delegations = delegations;
    await emitPhase(context, {
      type: "wave_complete",
      waveNumber: context.waveNumber,
    });

    return StageResult.Continue();
  }

  private buildTaskPrompt(task: any): string {
    return `
You are a CRAFTER agent. Your role is to implement the following task:

# ${task.title}

## Objective
${task.objective}

## Scope
${task.scope}

## Acceptance Criteria
${task.acceptanceCriteria.map((c: string) => `- ${c}`).join("\n")}

## Verification Commands
${task.verificationCommands.map((c: string) => `- ${c}`).join("\n")}

Please implement this task and ensure all acceptance criteria are met.
`.trim();
  }
}

